scylla/src/transport/load_balancing/default.rs

```diff
@@ -120,7 +122,24 @@ or refrain from preferring datacenters (which may ban all other datacenters, if
 
         // If token is available, get a shuffled list of alive replicas.
         let maybe_replicas = if let Some(ts) = &routing_info.token_with_strategy {
-            let local_replicas = self.shuffled_replicas(ts, true, Self::is_alive, cluster);
+            let local_rack_replicas = self.replicas(
+                ts,
+                true,
+                |node| {
+                    Self::is_alive(node)
+                        && DefaultPolicy::rack_predicate(node, &self.preferred_rack)
+                },
+                cluster,
+            );
+            let local_replicas = self.shuffled_replicas(
+                ts,
+                true,
+                |node| {
+                    Self::is_alive(node)
+                        && !DefaultPolicy::rack_predicate(node, &self.preferred_rack)
+                },
+                cluster,
+            );
```

Comment:
Judging from your `fallback()` implementation, you suggest such a load-balancing order:
1. local-rack local-DC replicas,
2. non-local-rack local-DC replicas,
3. remote replicas,
4. local nodes,
5. remote nodes.

As you haven't modified `pick()`, however, it chooses nodes in such an order:
1. local-rack local-DC replicas (because only then `pick_predicate() is true),
2. all other nodes.

It is probable that there are no replicas in the preferred rack. What then? I believe we should still prioritise local DC replicas, which currently is not true.